application_store/db/queries/application/queries.py

```diff
@@ -266,25 +269,96 @@ def submit_application(application_id) -> Applications:
+                            changed_fields.add(field["key"])
+                            if "history_log" in field:
+                                field["history_log"].append(
+                                    {datetime.now(tz=timezone.utc).isoformat(): field_map[field["key"]]}
+                                )
+                            else:
+                                field["history_log"] = [
+                                    {datetime.now(tz=timezone.utc).isoformat(): field_map[field["key"]]}
+                                ]
+
+            if changed_fields:
+                row["workflow_status"] = WorkflowStatus.CHANGE_RECEIVED
+
+            stmt = postgres_insert(AssessmentRecord).values(row)
+
+            # setting with an update makes sure the derived values are recalculated
+            update_row_statement = stmt.on_conflict_do_update(
+                index_elements=[AssessmentRecord.application_id], set_=row
+            ).returning(AssessmentRecord.application_id)
+
+            db.session.execute(update_row_statement)
+
+            flags = db.session.scalars(
+                select(AssessmentFlag).where(
+                    AssessmentFlag.application_id == row["application_id"],
+                    AssessmentFlag.is_change_request == True,  # noqa: E712
+                )
+            ).all()
+
+            for flag in flags:
+                if flag.field_ids and all(field_id in changed_fields for field_id in flag.field_ids):
+                    flag_update = FlagUpdate(
+                        justification="Applicant updated their submission",
+                        status=FlagStatus.RESOLVED,
+                        assessment_flag_id=flag.id,
+                    )
+                    flag.updates.append(flag_update)
+                    flag.latest_status = flag_update.status
+                    db.session.add(flag)
+                else:
+                    # Currently it shouldn't be possible to resubmit without addressing all the requested changes
+                    current_app.logger.warning(
+                        "Application has been resubmitted without resolving all change requests: {app_id}",
+                        extra=dict(app_id=row["application_id"]),
+                    )
+
+            db.session.commit()
         else:
-            current_app.logger.info(
-                "Successfully inserted application: {app_id}", extra=dict(app_id=row["application_id"])
+            stmt = postgres_insert(AssessmentRecord).values([row])
+
+            upsert_rows_stmt = stmt.on_conflict_do_nothing(index_elements=[AssessmentRecord.application_id]).returning(
+                AssessmentRecord.application_id
             )
-        db.session.commit()
+
+            result = db.session.execute(upsert_rows_stmt)
+
+            # Check if the inserted application is in result
+            inserted_application_ids = [item.application_id for item in result]
```

Comment:
Not sure what this achieves. Isn't the if below only len(result)?

Reply:
Yep - this was from the existing code but I didn't notice it; updated it now.

Reply:
🤯